sound/soc/sof/ipc4-telemetry.c

```diff
@@ -0,0 +1,114 @@
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2018-2023 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry.h"
+#include "ipc4-priv.h"
+
+static void *sof_ipc4_query_exception_address(struct snd_sof_dev *sdev, void *cached_addr)
+{
+	u32 telemetry_slot_offset;
+	void __iomem *io_addr;
+	void *start_addr;
+	u32 offset;
+
+	telemetry_slot_offset =
+		sof_ipc4_find_debug_slot_offset_by_type(sdev, cached_addr,
+							SOF_IPC4_DEBUG_SLOT_TELEMETRY);
+	if (!telemetry_slot_offset)
+		return NULL;
+
+	/* skip the first separator code */
+	offset = telemetry_slot_offset + sizeof(u32);
+
+	if (!pm_runtime_active(sdev->dev)) {
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE)
+		start_addr = cached_addr + offset;
+		return start_addr;
+#else
+		return NULL;
+#endif
+	}
+
+	io_addr = sdev->bar[sdev->mailbox_bar] + offset;
+	memcpy_fromio(cached_addr, io_addr, SOF_IPC4_DEBUG_SLOT_SIZE - 4);
+
+	return cached_addr;
+}
+
+static ssize_t sof_telemetry_entry_read(struct file *file, char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	loff_t pos = *ppos;
+	size_t size_ret;
+	void *start_addr;
+
+	if (pos < 0)
+		return -EINVAL;
+	if (pos >= SOF_IPC4_DEBUG_SLOT_SIZE - 4 || !count)
+		return 0;
+	if (count > SOF_IPC4_DEBUG_SLOT_SIZE - 4 - pos)
+		count = SOF_IPC4_DEBUG_SLOT_SIZE - 4 - pos;
+
+	start_addr = sof_ipc4_query_exception_address(sdev, dfse->cache_buf);
+	if (!start_addr)
+		return -EFAULT;
+
+	size_ret = copy_to_user(buffer, start_addr + pos, count);
+
+	/* update count & position if copy succeeded */
+	if (size_ret)
+		return -EFAULT;
+
+	*ppos = pos + count;
+
+	return count;
+}
+
+static const struct file_operations sof_telemetry_fops = {
+	.open = simple_open,
+	.read = sof_telemetry_entry_read,
+};
+
+void sof_ipc4_create_exception_debugfs_node(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_dfsentry *dfse;
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		return;
+
+	dfse->type = SOF_DFSENTRY_TYPE_IOMEM;
+	/* save the full size of debug_box to check telemetry slot in d3 case */
+	dfse->size = SOF_IPC4_DEBUG_SLOT_SIZE * (SOF_IPC4_MAX_DEBUG_SLOTS + 1);
+	dfse->io_mem = sdev->bar[sdev->mailbox_bar] + sdev->debug_box.offset;
+	dfse->access_type = SOF_DEBUGFS_ACCESS_D0_ONLY;
+	dfse->sdev = sdev;
+
+	dfse->cache_buf = devm_kzalloc(sdev->dev, dfse->size, GFP_KERNEL);
+	if (!dfse->cache_buf)
+		return;
+
+	list_add(&dfse->list, &sdev->dfsentry_list);
+
+	debugfs_create_file("exception", 0444, sdev->debugfs_root, dfse, &sof_telemetry_fops);
+}
+
+void sof_ipc4_handle_exception(struct snd_sof_dev *sdev)
+{
+	snd_sof_handle_fw_exception(sdev, "Fw Panic");
```

Comment:
This may keep the DSP active, so not sure if all those pm_runtime_active() checks have any merit.

````
void snd_sof_handle_fw_exception(struct snd_sof_dev *sdev, const char *msg)
{
	if (IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT) ||
	    sof_debug_check_flag(SOF_DBG_RETAIN_CTX)) {
		/* should we prevent DSP entering D3 ? */
		if (!sdev->ipc_dump_printed)
			dev_info(sdev->dev,
				 "Attempting to prevent DSP from entering D3 state to preserve context\n");
		pm_runtime_get_if_in_use(sdev->dev);
	}

	/* dump vital information to the logs */
	snd_sof_ipc_dump(sdev);
	snd_sof_dsp_dbg_dump(sdev, msg, SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX);
	sof_fw_trace_fw_crashed(sdev);
}

````

Reply:
Good hint! I didn't set CONFIG_SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT and SOF_DBG_RETAIN_CTX so I got d3 issue.  I also looking for this. I prefer to set SOF_DBG_RETAIN_CTX for fw panic case then we don't have trouble for accessing memory windows when hardware is in d3.